Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm

```diff
@@ -0,0 +1,311 @@
+
+    if (parameters.parentIdentifier) {
+        if (RefPtr parentMenuItem = m_extensionContext->menuItem(parameters.parentIdentifier.value()))
+            parentMenuItem->addSubmenuItem(*this);
+    }
+
+    if (parameters.contexts && !parameters.contexts.value().isEmpty())
+        m_contexts = parameters.contexts.value();
+    else if (m_parentMenuItem)
+        m_contexts = m_parentMenuItem->contexts();
+    else
+        m_contexts = WebExtensionMenuItemContextType::Page;
+
+    if (!parameters.iconDictionaryJSON.isEmpty())
+        m_icons = parseJSON(parameters.iconDictionaryJSON);
+}
+
+WebExtensionMenuItemParameters WebExtensionMenuItem::minimalParameters() const
+{
+    return {
+        identifier(),
+        parentMenuItem() ? std::optional { parentMenuItem()->identifier() } : std::nullopt,
+
+        type(),
+
+        nullString(), // title
+        nullString(), // command
+        nullString(), // iconDictionaryJSON
+
+        isChecked(),
+        isEnabled(),
+        isVisible(),
+
+        std::nullopt, // documentURLPatterns
+        std::nullopt, // targetURLPatterns
+
+        contexts(),
+    };
+}
+
+void WebExtensionMenuItem::update(const WebExtensionMenuItemParameters& parameters)
+{
+    ASSERT(extensionContext());
+
+    if (parameters.type)
+        m_type = parameters.type.value();
+
+    if (!parameters.identifier.isEmpty())
+        m_identifier = parameters.identifier;
+
+    if (parameters.parentIdentifier) {
+        RefPtr newParentMenuItem = m_extensionContext->menuItem(parameters.parentIdentifier.value());
+        if (newParentMenuItem.get() != m_parentMenuItem) {
+            if (m_parentMenuItem)
+                m_parentMenuItem->removeSubmenuItem(*this);
+
+            if (newParentMenuItem)
+                newParentMenuItem->addSubmenuItem(*this);
+        }
+    }
+
+    if (!parameters.title.isEmpty())
+        m_title = removeAmpersands(parameters.title);
+
+    if (!parameters.command.isNull())
+        m_command = extensionContext()->command(parameters.command);
+
+    if (!parameters.iconDictionaryJSON.isNull())
+        m_icons = parseJSON(parameters.iconDictionaryJSON);
+
+    if (parameters.checked)
+        m_checked = parameters.checked.value();
+
+    if (parameters.enabled)
+        m_enabled = parameters.enabled.value();
+
+    if (parameters.visible)
+        m_visible = parameters.visible.value();
+
+    if (parameters.contexts) {
+        if (!parameters.contexts.value().isEmpty())
+            m_contexts = parameters.contexts.value();
+        else if (m_parentMenuItem)
+            m_contexts = m_parentMenuItem->contexts();
+        else
+            m_contexts = WebExtensionMenuItemContextType::Page;
+    }
+}
+
+NSArray *WebExtensionMenuItem::matchingPlatformMenuItems(const MenuItemVector& menuItems, const WebExtensionMenuItemContextParameters& contextParameters, size_t limit)
+{
+    bool inRadioGroup = false;
+    bool groupChecked = false;
+    size_t count = 0;
```

Comment:
Does this need `__block`? Or is that only for Objective-C blocks?

Reply:
Only ObjC blocks, the compiler complains otherwise. The C++ lambda captures this as a reference with the `[&]` below.